File: cfassist-go/internal/memory/memory.go
package memory

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/aachtenberg/cfoperator/cfassist-go/internal/client"
)

// SaveConversation writes messages to a timestamped JSONL file.
func SaveConversation(memoryDir string, messages []client.Message) (string, error) {
	if len(messages) == 0 {
		return "", nil
	}

	if err := os.MkdirAll(memoryDir, 0755); err != nil {
		return "", err
	}

	// Build filename from timestamp + slug of first user message
	ts := time.Now().Format("2006-01-02T15-04-05")
	slug := ""
	for _, m := range messages {
		if m.Role == "user" {
			slug = slugify(m.Content, 40)
			break
		}
	}
	if slug == "" {
		slug = "conversation"
	}

	filename := fmt.Sprintf("%s_%s.jsonl", ts, slug)
	path := filepath.Join(memoryDir, filename)

	f, err := os.Create(path)
	if err != nil {
		return "", err
	}
	defer f.Close()

	enc := json.NewEncoder(f)
	for _, m := range messages {
		if err := enc.Encode(m); err != nil {
			return path, err
		}
	}
	return path, nil
}

// LoadConversation reads messages from a JSONL file.
func LoadConversation(filepath string) ([]client.Message, error) {
	f, err := os.Open(filepath)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	var messages []client.Message
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var m client.Message
		if err := json.Unmarshal([]byte(line), &m); err != nil {
			continue
		}
		messages = append(messages, m)
	}
	return messages, scanner.Err()
}

// Cleanup removes oldest conversations if over the limit.
func Cleanup(memoryDir string, maxConversations int) error {
	matches, err := filepath.Glob(filepath.Join(memoryDir, "*.jsonl"))
	if err != nil {
		return err
	}

	if len(matches) <= maxConversations {
		return nil
	}

	sort.Strings(matches)
	toDelete := len(matches) - maxConversations
	for i := 0; i < toDelete; i++ {
		os.Remove(matches[i])
	}
	return nil
}

var nonWordRe = regexp.MustCompile(`[^\w\s-]`)
var whitespaceRe = regexp.MustCompile(`[\s_]+`)

func slugify(text string, maxLen int) string {
	s := nonWordRe.ReplaceAllString(text, "")
	s = strings.ToLower(s)
	s = whitespaceRe.ReplaceAllString(s, "-")
	if len(s) > maxLen {
		s = s[:maxLen]
	}
	s = strings.TrimRight(s, "-")
	return s
}

File: cfassist/cfassist/memory.py
"""Conversation memory — save/load as JSONL files."""

import json
import re
from datetime import datetime
from pathlib import Path


def _slugify(text, max_len=40):
    """Create a filename-safe slug from text."""
    slug = re.sub(r"[^\w\s-]", "", text.lower().strip())
    slug = re.sub(r"[\s_]+", "-", slug)
    return slug[:max_len].rstrip("-")


def save_conversation(memory_dir, messages):
    """Save a conversation to a JSONL file.

    File is named: YYYY-MM-DDTHH-MM-SS_slug-of-first-message.jsonl
    """
    if not messages:
        return None

    dir_path = Path(memory_dir)
    dir_path.mkdir(parents=True, exist_ok=True)

    # Build filename from timestamp and first user message
    timestamp = datetime.now().strftime("%Y-%m-%dT%H-%M-%S")
    first_user = next((m["content"] for m in messages if m["role"] == "user"), "conversation")
    slug = _slugify(first_user)
    filename = f"{timestamp}_{slug}.jsonl"
    filepath = dir_path / filename

    with open(filepath, "w") as f:
        for msg in messages:
            entry = {
                "role": msg["role"],
                "content": msg.get("content", ""),
            }
            if "tool_calls" in msg:
                entry["tool_calls"] = msg["tool_calls"]
            f.write(json.dumps(entry, default=str) + "\n")

    return filepath


def load_conversation(filepath):
    """Load a conversation from a JSONL file."""
    messages = []
    with open(filepath) as f:
        for line in f:
            line = line.strip()
            if line:
                messages.append(json.loads(line))
    return messages


def list_conversations(memory_dir, limit=20):
    """List recent conversations, newest first."""
    dir_path = Path(memory_dir)
    if not dir_path.exists():
        return []

    files = sorted(dir_path.glob("*.jsonl"), reverse=True)
    results = []
    for f in files[:limit]:
        # Extract info from filename
        name = f.stem
        parts = name.split("_", 1)
        timestamp = parts[0] if parts else ""
        slug = parts[1] if len(parts) > 1 else ""
        results.append({
            "path": str(f),
            "timestamp": timestamp,
            "slug": slug,
        })
    return results


def cleanup_old_conversations(memory_dir, max_conversations=50):
    """Remove oldest conversations if over the limit."""
    dir_path = Path(memory_dir)
    if not dir_path.exists():
        return

    files = sorted(dir_path.glob("*.jsonl"))
    if len(files) <= max_conversations:
        return

    to_remove = files[: len(files) - max_conversations]
    for f in to_remove:
        f.unlink()
